File: platform-demos/C/samples/progressbar.py
from gi.repository import GLib
from gi.repository import Gtk
import sys


class MyWindow(Gtk.ApplicationWindow):
    # a window

    def __init__(self, app):
        Gtk.Window.__init__(self, title="ProgressBar Example", application=app)
        self.set_default_size(220, 20)

        # a progressbar
        self.progress_bar = Gtk.ProgressBar()
        # add the progressbar to the window
        self.add(self.progress_bar)

        # the method self.pulse is called each 100 milliseconds
        # and self.source_id is set to be the ID of the event source
        # (i.e. the bar changes position every 100 milliseconds)
        self.source_id = GLib.timeout_add(100, self.pulse)

    # event handler
    # any signal from the keyboard controls if the progressbar stops/starts
    def do_key_press_event(self, event):
        # if the progressbar has been stopped (therefore source_id == 0 - see
        # "else" below), turn it back on
        if (self.source_id == 0):
            self.source_id = GLib.timeout_add(100, self.pulse)
        # if the bar is moving, remove the source with the ID of source_id
        # from the main context (stop the bar) and set the source_id to 0
        else:
            GLib.source_remove(self.source_id)
            self.source_id = 0
        # stop the signal emission
        return True

    # source function
    # the progressbar is in "activity mode" when this method is called
    def pulse(self):
        self.progress_bar.pulse()
        # call the function again
        return True


class MyApplication(Gtk.Application):

    def __init__(self):
        Gtk.Application.__init__(self)

    def do_activate(self):
        win = MyWindow(self)
        win.show_all()

    def do_startup(self):
        Gtk.Application.do_startup(self)

app = MyApplication()
exit_status = app.run(sys.argv)
sys.exit(exit_status)

File: beginners-docs/C/samples/progressbar.js
#!/usr/bin/gjs

const Gio = imports.gi.Gio;
const GLib = imports.gi.GLib;
const Gtk = imports.gi.Gtk;
const Lang = imports.lang;

const ProgressBarExample = new Lang.Class({
    Name: 'ProgressBar Example',

    // Create the application itself
        _init: function() {
        this.application = new Gtk.Application({
            application_id: 'org.example.jsprogressbar',
            flags: Gio.ApplicationFlags.FLAGS_NONE
        });

    // Connect 'activate' and 'startup' signals to the callback functions
    this.application.connect('activate', Lang.bind(this, this._onActivate));
    this.application.connect('startup', Lang.bind(this, this._onStartup));
    },

    // Callback function for 'activate' signal presents windows when active
    _onActivate: function() {
        this._window.present();
    },

    // Callback function for 'startup' signal builds the UI
    _onStartup: function() {
        this._buildUI ();
    },



    // Build the application's UI
    _buildUI: function() {

        // Create the application window
            this._window = new Gtk.ApplicationWindow({ application: this.application,
                                                       window_position: Gtk.WindowPosition.CENTER,
                                                       default_height: 20,
                                                       default_width: 220,
                                                       title: "ProgressBar Example"});

        // Create the progress bar
        this.progressBar = new Gtk.ProgressBar ();
        this._window.add(this.progressBar);

        // Start the function that pulses the bar every 100 milliseconds
        this.sourceID = GLib.timeout_add(GLib.PRIORITY_DEFAULT, 100, Lang.bind(this, this._barPulse));

        // Connect a keypress event to the function that toggles the bar to start or stop pulsing
        this._window.connect("key-press-event", Lang.bind(this, this._onKeyPress));

            // Show the window and all child widgets
            this._window.show_all();
    },



    // Pulse the progressbar (unless it has been disabled by a keypress)
    _barPulse: function() {
        this.progressBar.pulse();
        return true;
    },

    // Start or stop the progressbar when a key is pressed
    _onKeyPress: function() {
        if (this.sourceID == 0)
            this.sourceID = GLib.timeout_add(GLib.PRIORITY_DEFAULT, 100, Lang.bind(this, this._barPulse));
        else {
            GLib.source_remove(this.sourceID);
            this.sourceID = 0;
        }
    }

});

// Run the application
let app = new ProgressBarExample ();
app.application.run (ARGV);
